torch/_subclasses/fake_tensor.py

```diff
@@ -227,6 +227,13 @@ def from_real_tensor(self, fake_mode, t, make_constant=False, shape_env=None):
             out.requires_grad_(t.requires_grad)
             if make_constant:
                 self.add_constant_storage_mapping(out)
+
+        # We need `is_leaf` metadata to be consistent
+        # See Note [Fake up some autograd history]
+        if t.requires_grad and not t.is_leaf:
+            with torch.enable_grad():
+                out = out.clone()
```

Comment:
Is this a fake tensor here

Reply:
yep, we're guaranteed it's a fake tensor here so no there's no worry of extra pointless compute.

side note - we need this logic in multiple places, so a `tensor._unsafe_set_grad_fn_field(...)` API that dangerously sets the `_grad_fn` on a tensor would be pretty nice. Then again, faking grad history isn't too hard anyway, and that API could be pretty easily mis-used, so w/e

Reply:
Do we also want to keep the autograd view propery?
We could also use something like `.alias()` if we want the created Tensor to have `._is_view() == True`.

Reply:
This is kind of annoying because `meta_converter` actually did all of this correctly, but the subsequent FakeTensor wrapping destroys the information that meta converter setup. Maybe a better way to do this is to just unify meta/fake converter onto the original class (or at the very least make the subclass instantiation configurable) so that we don't have to be manually reimplementing all of the logic here.

Reply:
In particular the view recreating logic is quite intricate and I'd prefer to only have one copy of it

Reply:
> Maybe a better way to do this is to just unify meta/fake converter onto the original class

Hmm could you explain what this looks like a bit more? It looks like today, we have:

(1) `MetaConverter`, which knows how to convert a plain old `torch.Tensor` to a meta tensor with all of the metadata edge cases properly handled, and does also handles caching
(2) `FakeTensorConverter`, which (among other things) knows how to convert a plain old `torch.Tensor` into a `FakeTensor`. It does this by first using `MetaConverter` to create a plain meta tensor, and then wrapping the meta tensor into a `FakeTensor` that we return. Constructing the `FakeTensor` requires calling `_make_subclass()`, which unfortunately doesn't reflect all of the metadata on the input tensor.

Is your suggestion just that `MetaConverter` and `FakeTensorConverter` need to re-use some of the same logic, so we should factor that out into a helper?

Reply:
You can do it that way, or you can make metaconverter directly accept the cls that its fresh tensors should be created from. This is how __new__ works in Python; it takes a cls so subclasses can use the constructor but still end up with a subclass

Reply:
this is actually kinda annoying... `MetaConverter` has a bunch of extra logic fused into it, that we don't actually want to run when we're creating a fake tensor. AND `MetaConverter.meta_tensor()` is recursive. "Other stuff" includes:

(1) caching. `MetaConverter` contains a `tensor_memo` of real-to-meta tensors. I don't think we want to pollute this with meta tensors?
(2) storage handling . `MetaConverter` does a bunch of stuff like creating a fresh torch.tensor to mock up a storage, and `.set()`'ing that storage + fresh sizes and strides on the newly created tensor. This is all totally unnecessary for fake tensor (since its storage is already set correctly), and feels like it could do the wrong thing.

Reply:
cc @ezyang - the refactor that I attempted was to re-use `MetaConverter.meta_tensor()`, and re-purpose it to accept a generic "constructor" (lambda). So `FakeTensor.from_real_tensor()` would call it twice - once to make a meta tensor from a real tensor, and again to make a FakeTensor from the meta tensor.

Agreed that this is not very clean, and caused a bunch of test failures that didn't finish tracking down (yet). Let me know if you want to look at this, or if you'd rather land the "simple" fix for is_leaf that doesn't involve a refactor.

Reply:
ok to land this fix, it is small and comes with a test

Reply:
Can you also check the test case in https://github.com/pytorch/pytorch/pull/87513/files

I tried to repro that from the original model.

Reply:
More general fix wip at https://github.com/pytorch/pytorch/pull/87943